src/cloudflare/internal/workflows-api.ts

```diff
@@ -1,12 +1,173 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
 export class NonRetryableError extends Error {
-  // `__brand` is needed for Workflows' engine to validate if the user returned a NonRetryableError
-  // this provides better DX because they can just extend NonRetryableError for their own Errors
-  // and override name.
-  // This needs to be a public field because it's serialized over RPC to the Workflows' engine
+  // `__brand` is needed for engine to validate if the user returned a NonRetryableError
+  // This provides a better DX because they can just extend NonRetryableError for their own Errors
+  // and override name
+  // Private fields are not serialized over RPC
   public readonly __brand: string = 'NonRetryableError';
 
   public constructor(message: string, name = 'NonRetryableError') {
     super(message);
     this.name = name;
   }
 }
+
+type InstanceStatus = {
+  status:
+    | 'queued'
+    | 'running'
+    | 'paused'
+    | 'errored'
+    | 'terminated'
+    | 'complete'
+    | 'unknown';
+  error?: string;
+  output?: object;
+};
+
+interface Fetcher {
+  fetch: typeof fetch;
+}
+
+async function callFetcher<T>(
+  fetcher: Fetcher,
+  path: string,
+  body: object
+): Promise<T> {
+  const response = await fetcher.fetch(`http://workflow-binding.local${path}`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      'X-Version': '1',
+    },
+    body: JSON.stringify(body),
+  });
+
+  // TODO: Throw if result is not 200 or ok is false
+
+  return response.json() as Promise<T>;
+}
+
+export class Instance {
+  private readonly fetcher: Fetcher;
+  private readonly instanceId: string;
+
+  public constructor(instanceId: string, fetcher: Fetcher) {
+    this.instanceId = instanceId;
+    this.fetcher = fetcher;
+  }
+
+  public async pause(): Promise<void> {
+    try {
+      await callFetcher(this.fetcher, '/pause-instance', {
+        instanceId: this.instanceId,
+      });
+    } catch (_error) {
+      throw new WorkflowError(
```

Comment:
do you want to add anything about the error here? Like status code at least. Otherwise debugging this will be hard.

Reply:
Yeah, this was a first pass. I think it would be nice to have more structured errors here perhaps. I'll give that a shot now.

Reply:
I've decided to let the wrapped binding Worker return back the error (since it has a lot more context) and just wrap that in a `WorkflowError ` here